### PR TITLE
PyTorch backend: Reduce FP16 overhead

### PR DESCRIPTION
**Pull request by tomtseng**
## Context
I found that FP16 with ViT-b4 and ViT-b8 were slower than FP32. The FP16 overhead was from `torch::Tensor::to()` converting tensors between FP32 (data type that KataGo uses inputs and outputs) and FP16 (data type for the input/output of a FP16 TorchScript model).

## Changes
In this PR I move the FP32 <-> FP16 conversion into `SymmetryHelpers::copy{Inputs,Outputs}WithSymmetry()`, which reduces the overhead (though there is still some overhead).

## Testing

* Match ViT-b4-fp16-v16 vs. ViT-b4-fp32-v16 vs. katago-b10-s197m-v64 and check win rates.
Before PR:
```
Wins by row player against column player:
                vit-fp32-v16  vit-fp16-v16  b10-s197m-v6
  vit-fp32-v16             -   212.0/400.0   228.0/400.0
  vit-fp16-v16   188.0/400.0             -   240.0/400.0
 b10-s197m-v64   172.0/400.0   160.0/400.0             -
Win% by row player against column player:
                vit-fp  vit-fp  b10-s1
  vit-fp32-v16       -   53.0%   57.0%
  vit-fp16-v16   47.0%       -   60.0%
 b10-s197m-v64   43.0%   40.0%       -
```
After PR:
```
Wins by row player against column player:
                vit-fp32-v16  vit-fp16-v16  b10-s197m-v6
  vit-fp32-v16             -   216.0/400.0   234.0/400.0
  vit-fp16-v16   184.0/400.0             -   236.0/400.0
 b10-s197m-v64   166.0/400.0   164.0/400.0             -
Win% by row player against column player:
                vit-fp  vit-fp  b10-s1
  vit-fp32-v16       -   54.0%   58.5%
  vit-fp16-v16   46.0%       -   59.0%
 b10-s197m-v64   41.5%   41.0%       -
 ```
 
 * Benchmark selfplay throughput and checked that it is much faster. (Oddly, match isn't much faster.)

Config for match:
```
numGamesTotal=1200
numBots=3
numGameThreads=256
nnMaxBatchSize=256

nnModelFile0=/nas/ucb/ttseng/go_attack/scratch/vit-b16-s650m/t0-s0-d0/model-fp16.pt
botName0=vit-fp16-v16
maxVisits0=16
searchAlgorithm0=MCTS
useGraphSearch0=true
useNHWC0=false
inputsUseNHWC0=false
useFP16-0=true
nnModelFile1=/nas/ucb/ttseng/go_attack/scratch/vit-b16-s650m/t0-s0-d0/model.pt
botName1=vit-fp32-v16
maxVisits1=16
searchAlgorithm1=MCTS
useGraphSearch1=true
useNHWC1=false
inputsUseNHWC1=false
useFP16-1=false

nnModelFile2=/nas/ucb/k8/go-attack/katago-networks/kata1-b10c128-s197428736-d67404019.txt.gz
botName2=b10-s197m-v64
maxVisits2=64
searchAlgorithm2=MCTS
useGraphSearch2=true
```

### CODE REVIEW

File: cpp/neuralnet/nninputs.h [hunk 1]
```diff
@@ -240,4 +243,81 @@ namespace ScoreValue {
   double getScoreStdev(double scoreMeanAvg, double scoreMeanSqAvg);
 }
 
+////////////////////
+// Implementation //
+////////////////////
+
+template<typename SrcT, typename DstT>
+static void copyWithSymmetry(const SrcT* src, DstT* dst, int nSize, int hSize, int wSize, int cSize, bool useNHWC, int symmetry, bool reverse) {
```
Comment (tomtseng) [inline cpp/neuralnet/nninputs.h:251]: The entire function body is the same as it was in `nninputs.cpp`, the only difference is the first two args are templated.

File: cpp/neuralnet/pytorchbackend.cpp [hunk 2]
```diff
@@ -292,9 +336,9 @@ void getOutput(
 
     if (output->whiteOwnerMap != NULL) {
       if (!ownershipOutputs.defined()) {
-        ownershipOutputs = modelOutputs[4].toTensor().to(at::kCPU, torch::kFloat32);
```
Comment (tomtseng) [inline cpp/neuralnet/pytorchbackend.cpp:295]: this line is pretty bad, it's moving the entire GPU tensor (for all rows) to CPU for each row in the batch

**Comment by tomtseng** [comment]
this has sat around for a while and has been used in ViT training run, I'm just going to merge and clear out all these stale PRs

**Comment by AdamGleave** [comment]
Should we get this merged? Is it just waiting @ed1d1a8d's review? @ed1d1a8d are you able to do this or should we reassign the review?

**Review by AdamGleave — APPROVED** [review]
Maybe LGTM

**Comment by tomtseng** [comment]
merging without comprehensive review:
* Tony doesn't have bandwidth
* Getting someone else to review outside this project wouldn't be worth it (it'd take them some time to get enough context, and I wouldn't expect the resulting review to be very deep)
* We've already run experiments using this (various evals) and it doesn't seem to have hurt playing strength

**Pull request merged**